opm/simulators/wells/BlackoilWellModelGeneric.cpp

```diff
@@ -82,6 +85,238 @@ namespace {
 
         return { 0.0, restart.get(item), target };
     }
+
+    struct RetrieveWellGuideRate
```

Comment:
Looks like this functionality could be tested without bringing in the full vanguard?

Reply:
Looks like the general group control code could benefit from something like this?

Reply:
> We do need a mostly complete Schedule object with a non-trivial group tree though

I think this should be brought out with a header file++ and then tested; could use e.g. `MOD_GRP.DATA` as test data?

Reply:
> > We do need a mostly complete Schedule object with a non-trivial group tree though
> 
> I think this should be brought out with a header file++ and then tested; could use e.g. `MOD_GRP.DATA` as test data?

On the other hand, there's just a single use-case at the moment and I'd rather not generalise from that single example.  For instance, we don't know how `GroupTreeWalker` will hold up if deployed to more general use-cases, especially ones that may involve communicating partial results between MPI ranks during group tree traversal.

In other words, I'd prefer to keep this completely hidden from other parts of the code until we can explore the problem more broadly.

Reply:
I beg to differ here:

1. We don't have any API stability in this project - at all; i.e. there is no real harm in exposing functionality which will probably change in the future. On the contrary it would be valuable to have *current* behavior and expectations codified in a test when it is at a later staged extended in new directions.
2. opm-simulators is notoriously difficult to test because many things are strongly coupled. This is an example of functionality which is not strongly coupled to other parts of the code (very good!) and therefor testing is relatively simpler, I really think we should use the opportunities to add tests to opm-simulators when the opportunity arises.

I have made my point; I'll back off now.

Reply:
> In other words, I'd prefer to keep this completely hidden from other parts of the code until we can explore the problem more broadly.

I disagree with this, refactoring to move the code with generalizing potential out of its hiding place in a cpp file is good! Otherwise it becomes very unlikely that it will never be improved or reused, since any interested such user first must extract it and make sure that was ok. It is also not necessarily good to put all implementation details in a cpp file, putting classes in their own file is an aid to understanding.

Reply:
I have another idea, that may be worth considering. We could implement the Visitor pattern for group and well hierarchies. I think that is well suited for our situation: a stable set of classes in the tree (groups and wells) and an ever-increasing set of operations we would like to apply/compute in the hierarchy. So virtual methods are a bad choice since we would need to add new virtual methods to the nodes all the time. New visitor classes do not require changes to the well and group classes.

Reply:
> > In other words, I'd prefer to keep this completely hidden from other parts of the code until we can explore the problem more broadly.
> 
> I disagree with this, refactoring to move the code with generalizing potential out of its hiding place in a cpp file is good! Otherwise it becomes very unlikely that it will never be improved or reused, since any interested such user first must extract it and make sure that was ok. It is also not necessarily good to put all implementation details in a cpp file, putting classes in their own file is an aid to understanding.

We need the overall feature of this PR (reporting the well-level injection and production guide rates to the summary file) rather soon, so I suggest we merge this PR in its current state.  Then I'll extract the `GroupTreeWalker` in a follow-up PR.  I'd still prefer not to do so until we agree on its (mostly) final location though, because I can envision using it in opm-common too.

> We could implement the Visitor pattern for group and well hierarchies

Yes, that's a good idea that would complement the current facility.  For accumulation and distribution purposes you still need to know if you visit the parent before the child (or children) or the other way around, but there could be multiple hierarchy traversal strategies.

Reply:
> We need the overall feature of this PR (reporting the well-level injection and production guide rates to the summary file) rather soon, so I suggest we merge this PR in its current state. Then I'll extract the `GroupTreeWalker` in a follow-up PR. I'd still prefer not to do so until we agree on its (mostly) final location though, because I can envision using it in opm-common too.

Ok, that I can get behind.

> For accumulation and distribution purposes you still need to know if you visit the parent before the child (or children) or the other way around, but there could be multiple hierarchy traversal strategies.

Typically all we would need for any visitor is is preVisit(Well), postVisit(Well), preVisit(Group) and postVisit(Group). We could also (orthogonally) support depth first and breadth first visit order.